core/src/main/java/org/apache/iceberg/OverwriteData.java

```diff
@@ -88,6 +119,42 @@ public OverwriteFiles validateAddedFiles() {
       }
     }
 
+    if (conflictDetectionFilter != null) {
+      PartitionSpec spec = writeSpec();
+      Expression inclusiveExpr = Projections.inclusive(spec).project(conflictDetectionFilter);
+      Evaluator inclusive = new Evaluator(spec.partitionType(), inclusiveExpr);
+
+      InclusiveMetricsEvaluator metrics = new InclusiveMetricsEvaluator(base.schema(), conflictDetectionFilter);
+
+      List<DataFile> newFiles = collectNewFiles(base);
+      for (DataFile newFile : newFiles) {
+        ValidationException.check(
+            !inclusive.eval(newFile.partition()) || !metrics.eval(newFile),
+            "A conflicting file was appended that matches filter '%s': %s",
```

Comment:
Minor: This isn't quite correct. This doesn't guarantee that a file matches the filter, just that it may contain data that matches the filter. For example, if a table is bucketed by `id` and a file has ids `x` and `z` where `x < y` and `y < z` and `x`, `y`, and `z` all hash to the same bucket, then this would match the file even though `y` isn't actually in it.

Reply:
You are right, that's exactly the behavior we wanted to achieve. If the file can potentially contain data that matches the filter, we should fail. I should update the error message.